cmd/pod-scaler/frontend/src/app/CircLLHist/LogarithmicComparativePlot.tsx

```diff
@@ -0,0 +1,403 @@
+  let largest = 0;
+  for (const bin of bins) {
+    if (bin.count > largest) {
+      largest = bin.count
+    }
+  }
+  mainContext.fillStyle = d3.interpolateCividis(0);
+  mainContext.globalCompositeOperation = "destination-over";
+  mainContext.fillRect(offset, canvasMeta.verticalOffset + canvasMeta.topOffset, canvasMeta.cellWidth, canvasMeta.dataHeight)
+  mainContext.globalCompositeOperation = "source-over";
+  for (const bin of bins) {
+    if (bin.count === 0) {
+      continue
+    }
+    const bottom = Value(bin);
+    const height = Width(bin);
+    const top: number = bottom + height;
+    if (bottom === 0 || height === 0 || bottom < canvasMeta.minimum || top < canvasMeta.minimum) {
+      continue
+    }
+    mainContext.fillStyle = d3.interpolateCividis(bin.count / largest)
+    const topY = pixelCoordinate(canvasMeta, top);
+    const bottomY = pixelCoordinate(canvasMeta, bottom);
+    const heightY = bottomY - topY;
+    mainContext.fillRect(offset, topY, canvasMeta.cellWidth, heightY)
+
+    const hiddenColor = nextColor(canvasMeta);
+    index[hiddenColor] = "(" + canvasMeta.yAxisFormatter(bottom) + "," + canvasMeta.yAxisFormatter(bottom + height) + ")" + canvasMeta.yAxisUnit + ": " + bin.count + " samples";
+    hiddenContext.fillStyle = hiddenColor
+    hiddenContext.fillRect(offset, topY, canvasMeta.cellWidth, heightY)
+  }
+}
+
+export const LogarithmicComparativePlot: React.FunctionComponent<LogarithmicComparativePlotProps> = (props: LogarithmicComparativePlotProps) => {
+  const container: React.RefObject<HTMLDivElement> = React.useRef<HTMLDivElement>(null);
+  const [canvasMeta, setCanvasMeta] = React.useState<CanvasMetadata>(canvasMetaFromProps(props, undefined));
+
+  // re-compute window on resize
+  React.useEffect(() => {
+    const handleResize = () => {
+      if (container.current) {
+        const width = container.current.getBoundingClientRect().width;
+        setCanvasMeta(canvasMetaFromProps(props, width))
+      }
+    };
+    window.addEventListener('resize', handleResize);
+    handleResize();
+    return () => {
+      window.removeEventListener('resize', handleResize);
+    };
+  }, [container, props]);
+
+  const [first, setFirst] = React.useState<number>(0);
+
+  const onChange = (value: number): void => {
+    setFirst(value);
+  };
+
+  const [index, setIndex] = React.useState<Record<string, string>>({});
+  const mainCanvas: React.RefObject<HTMLCanvasElement> = React.useRef<HTMLCanvasElement>(null);
+  const hiddenCanvas: React.RefObject<HTMLCanvasElement> = React.useRef<HTMLCanvasElement>(null);
+  const tooltip: React.RefObject<HTMLDivElement> = React.useRef<HTMLDivElement>(null);
+
+  // recompute meta when data changes
+  React.useEffect(() => {
+    if (container.current) {
+      const width = container.current.getBoundingClientRect().width;
+      setCanvasMeta(canvasMetaFromProps(props, width))
+    }
+  }, [props])
```

Comment:
I believe this will effect hook will re-rerun on every render. I'd expect `props` to be a different object every time.

Reply:
If any field in `props` changes, I need to re-draw - how do I express that? I don't care if the container object for the fields changes if the data inside does not.

Reply:
One option would be to leave off the dependency list so the effect hook runs every render, then use `React.memo` for your component.

https://reactjs.org/docs/react-api.html#reactmemo

Otherwise, you could use `[props.a, props.b, props.c, ...]` as your dependencies.

I could be wrong here, though... You might want to check whether `props` is a different object each render.